src/v/kafka/server/handlers/alter_partition_reassignments.cc

```diff
@@ -0,0 +1,283 @@
+                      *unkown_broker_id_it,
+                      all_node_ids);
+                    resp.data.error_code
+                      = error_code::invalid_replica_assignment;
+                    resp.data.error_message = fmt::format(
+                      "Replica assignment has brokers that are not alive. "
+                      "Replica list: {}, live broker list: {}",
+                      *partition.replicas,
+                      all_node_ids);
+                    co_return co_await ctx.respond(std::move(resp));
+                }
+
+                auto assignment_exists
+                  = co_await ctx.topics_frontend().replica_set_exists(
+                    ntp, *partition.replicas);
+                if (assignment_exists) {
+                    // The requested reassignment already exists, so skip this
+                    // one
+                    vlog(
+                      klog.info,
+                      "Request to change ntp {} replica set to {}, no change",
+                      ntp,
+                      *partition.replicas);
+                    continue;
+                }
+
+                // Convert each node id to a broker shard because
+                // topics_frontend.move_partition_replicas expects
+                // vector<broker_shard>
+                std::vector<model::broker_shard> replicas;
+                std::for_each(
+                  partition.replicas->cbegin(),
+                  partition.replicas->cend(),
+                  [&replicas](const model::node_id& node_id) {
+                      replicas.push_back(
+                        model::broker_shard{.node_id = node_id, .shard = 0});
+                  });
+
+                vlog(
+                  klog.debug,
+                  "Request to reassign partitions: ntp {}, replicas {}",
+                  ntp,
+                  replicas);
+
+                auto errc
+                  = co_await ctx.topics_frontend().move_partition_replicas(
+                    ntp, replicas, model::timeout_clock::now() + timeout_ms);
+
+                if (!errc) {
+                    topic_response.partitions.push_back(
+                      reassignable_partition_response{
+                        .partition_index = partition.partition_index});
+                } else {
+                    vlog(
+                      klog.error,
+                      "Unknown problem moving partitions: ntp {}, ec {}",
+                      ntp,
+                      errc);
+                    auto kerr = cluster_errc_to_kafka_error(errc);
+                    topic_response.partitions.push_back(
+                      reassignable_partition_response{
+                        .error_code = kerr,
+                        .error_message = ss::sstring{error_code_to_str(kerr)}});
+                }
+
+            } else {
+                // Otherwise we cancel the pending request
+                auto errc = co_await ctx.topics_frontend()
```

Comment:
This seems strange that AlterPartitionReassignments request could cancel pending work, is this how its done in kafka? I checked KIP-455, theres no mention of that in there.

Reply:
Canceling pending work is mentioned all over KIP-455, but the first place is in the schemata for AlterPartitionsReassignmentsRequest:
```
        { "name": "Replicas", "type": "[]int32", "versions": "0+", "nullableVersions": "0+", "default": "null",
          "about": "The replicas to place the partitions on, or null to cancel a pending reassignment for this partition." }
```

Also, apache does this.
They use a null value to represent cancelling a pending request:
The null value is created in KafkaApis.scala:
```scala
    val reassignments = alterPartitionReassignmentsRequest.data.topics.asScala.flatMap {
      reassignableTopic => reassignableTopic.partitions.asScala.map {
        reassignablePartition =>
          val tp = new TopicPartition(reassignableTopic.name, reassignablePartition.partitionIndex)
          if (reassignablePartition.replicas == null)
            tp -> None // revert call <-- this represents a cancel on a pending request
          else
            tp -> Some(reassignablePartition.replicas.asScala.map(_.toInt))
      }
    }.toMap
```

Then they pass that to the controller
```scala
  /**
   * Process a partition reassignment from the AlterPartitionReassignment API. If there is an
   * existing reassignment through zookeeper for any of the requested partitions, they will be
   * cancelled prior to beginning the new reassignment. Any zk-based reassignment for partitions
   * which are NOT included in this call will not be affected.
   *
   * @param reassignments Map of reassignments passed through the AlterReassignments API. A null value
   *                      means that we should cancel an in-progress reassignment.
   * @param callback Callback to send AlterReassignments response
   */
private def processApiPartitionReassignment
```

Reply:
Awesome, @NyaliaLui. The source code takes precedence over stuff in the KIP unless it looks like a bug :)

Reply:
Nice, thanks !